packages/gatsby/src/utils/worker/render-html.ts

```diff
@@ -1,50 +1,315 @@
-import Promise from "bluebird"
+import Bluebird from "bluebird"
 import { join } from "path"
+import { get, isObject, flatten, uniqBy, concat } from "lodash"
+
 import { getPageHtmlFilePath } from "../../utils/page-html"
+import { IPageDataWithQueryResult } from "../../utils/page-data"
 import { Stage } from "../../commands/types"
 
-export const renderHTML = ({
+let lastRenderHTMLCallTimestamp = 0
+let htmlComponentRenderer
+let webpackStats
+
+const staticQueryResultCache = new Map<string, any>()
+const inFlightStaticQueryPromise = new Map<string, Promise<any>>()
+
+const inlineCssCache = new Map<string, string>()
+const inFlightInlineCssPromise = new Map<string, Promise<string>>()
+
+const resourcesForTemplateCache = new Map<string, IResourcesForTemplate>()
+const inFlightResourcesForTemplate = new Map<
+  string,
+  Promise<IResourcesForTemplate>
+>()
+
+const getStaticQueryPath = (hash: string): string =>
+  join(`page-data`, `sq`, `d`, `${hash}.json`)
+
+const getStaticQueryResult = async (hash: string): any => {
+  const staticQueryPath = getStaticQueryPath(hash)
+  const absoluteStaticQueryPath = join(process.cwd(), `public`, staticQueryPath)
+  const staticQueryRaw = await fs.readFile(absoluteStaticQueryPath)
+
+  return JSON.parse(staticQueryRaw.toString())
+}
+
+async function readPageData(
+  publicDir: string,
+  pagePath: string
+): Promise<IPageDataWithQueryResult> {
+  const fixedPagePath = pagePath === `/` ? `index` : pagePath
+  const filePath = join(publicDir, `page-data`, fixedPagePath, `page-data.json`)
+  const rawPageData = await fs.readFile(filePath, `utf-8`)
+
+  return JSON.parse(rawPageData)
+}
+
+async function readWebpackStats(publicDir: string): any {
+  const filePath = join(publicDir, `webpack.stats.json`)
+  const rawPageData = await fs.readFile(filePath, `utf-8`)
+
+  return JSON.parse(rawPageData)
+}
+
+interface IScriptsAndStyles {
+  scripts: Array<any>
+  styles: Array<any>
+  reversedStyles: Array<any>
+  reversedScripts: Array<any>
+}
+
+async function getScriptsAndStylesForTemplate(
+  componentChunkName
+): Promise<IScriptsAndStyles> {
+  // Create paths to scripts
+  let scriptsAndStyles = flatten(
+    [`app`, componentChunkName].map(s => {
+      const fetchKey = `assetsByChunkName[${s}]`
+
+      let chunks = get(webpackStats, fetchKey)
+      const namedChunkGroups = get(webpackStats, `namedChunkGroups`)
+
+      if (!chunks) {
+        return null
+      }
+
+      chunks = chunks.map(chunk => {
+        if (chunk === `/`) {
+          return null
+        }
+        return { rel: `preload`, name: chunk }
+      })
+
+      namedChunkGroups[s].assets.forEach(asset =>
+        chunks.push({ rel: `preload`, name: asset })
+      )
+
+      const childAssets = namedChunkGroups[s].childAssets
+      for (const rel in childAssets) {
+        chunks = concat(
+          chunks,
+          childAssets[rel].map(chunk => {
+            return { rel, name: chunk }
+          })
+        )
+      }
```

Comment:
Nit: to build this final `chunks` array we first use `map`, then `forEach` with `push` and then `for` with `concat` 😄 Maybe prettify for consistency?

I realize this was originally this way in `static-entry.js` but I doubt we'll find a better time to refactor it a bit :)

Reply:
yeah, this was straight c&p, but we can use this opportunity to streamline this code (if not now then we never will heh)

Reply:
TBH, I'll probably not use any "functional" array methods and just init array(s) ahead of time and do `for` loops adding as many things as needed inside.

Reply:
done in https://github.com/gatsbyjs/gatsby/pull/29396/commits/1e87c4127e97023eab50eac7ae5294d138ab47c0 + follow up fix for 1 thing I messed up in https://github.com/gatsbyjs/gatsby/pull/29396/commits/7c46317e957fd48bebf5348cb74286a4ceb2a342